rfcs/0039-minaform-tool-design.md

```diff
@@ -0,0 +1,117 @@
+
+Additional flags can be provided to the create-network command for changing the quantities of fish, whales, seeds, bots, etc.
+
+The flag `--generate-image` will cause the tool to build a new image based on the provided one, but with the genesis ledger, proofs, and keys baked-in to reduce network startup time at the expense of this one-time step (Pain Point #4).
+
+Two flags are availible to specify which version of the daemon to run, `--daemon-image` and `--daemon-git-sha` where daemon-git-sha calculates the image tag based on the given git tag (Pain Point #5). 
+
+Once the network has been defined and set up, you can make manual changes to the terraform files to customize the new network, or just proceed to `minaform start <testnet_name>`.
+
+### Minaform Start
+
+The start command handles deploying the network and starting all containers / gcloud resources. `minaform start <testnet_name>` will start the given testnet on the default cluster, and provide a output to display the deployment progress. Keys will be uploaded to the cluster when the namespace is created, and then terraform resources will be deployed. If `--wait` is specified, the script will run `minaform watch <testnet_name>` once terraform is complete to watch for all of the nodes to become ready.
+
+### Minaform Stop
+
+The stop command handles destroying all resources for the given testnet, ideally using the existing terraform state and all of the same pararemeters so destroys can always happen smoothly. Will provide a `--force` flag to force terraform and other tooling to delete everything and clean up. This command closely matching the process for start/restart/etc. addresses pain point #3 and should make it easier to keep the cluster clean and easy.
+
+### Minaform Restart
+
+The restart command runs a `minaform stop` followed by `minaform start` and accepts the same flags as both start + stop.
+
+### Minaform Upgrade
+
+The upgrade command runs a terraform apply, to update the existing deployment with new changes. Ideally this should also provide more debugging flags to terraform and other tooling, and require user input to proceed so that the engineer deploying can understand what state will be lost (if any) in the process. Accepts the same flags as start/stop/restart.
+
+### Minaform Watch
+
+Addressing Pain Point #2 is the `minaform watch` command that will integrate with any scripts or tools we have right now (and write going forward) for monitoring network health. For grafana and google cloud it will just print links to their web dashboards. For kubernetes startup/liveness/readiness probes and tracking pod health there will be live statistics printed to the console. When new tooling is created for monitoring network health it should be added at least as a flag / subcommand of watch to allow for easy visibility.
```

Comment:
I don't think it makes sense to push this logic into the same script that is managing the deployments. The logic for monitoring the network is different enough and can involve enough different strategies that I don't think a monolithic script is the best way to actually maintain these kind of checks. I agree with the pain points around how we do status checks atm, but I personally believe we should split that out into a separate conversation from this tool.

Reply:
Agreed and, considering that the primary if not only datasources to watch for observing network health are Prometheus metrics (+ Grafana dashboards) and stackdriver logs, in my opinion consolidating logs and metrics into a single tool (Grafana in this case) could go a long way to allowing for easy visibility. Ideally the deployment tooling, however constructed, could just spit out a URL to a templated Grafana dashboard or group of dashboards scoped to the testnet being launched. 

To a previous point re: removing grafana labs from the equation, dedicated prometheus and grafana instances could help here and would align with the goal of ensuring proper cleanup of testnets since it would all be managed within the testnet terraform module.